src/mpid/ch4/src/mpidig_part.h

```diff
@@ -88,22 +81,40 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_parrived(MPIR_Request * request, int par
     int mpi_errno = MPI_SUCCESS;
     MPIR_FUNC_ENTER;
 
-    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
-
-    /* Do not maintain per-partition completion. Arrived when full data transfer is done.
-     * An inactive request returns TRUE (same for NULL req, handled at MPIR layer). */
-    if (MPIR_Request_is_complete(request)) {
+    MPIR_Assert(MPIR_Part_request_is_active(request));
+    MPIR_Assert(request->kind == MPIR_REQUEST_KIND__PART_RECV);
+    const MPIR_cc_t *cc_part = MPIDIG_PART_REQUEST(request, u.recv.cc_part);
+
+    // if the request has not matched or it's too early
+    const bool too_early = cc_part == NULL || !MPIR_Part_request_is_active(request);
+
+    // get the span of send request to check from start_part to end_part (included!)
+    const int send_part = too_early ? (0) : MPIDIG_PART_REQUEST(request, u.recv.send_part);
+    const int recv_part = request->u.part.partitions;
+    const int start_part = (partition * send_part) / recv_part;
+    const int n_part = send_part / recv_part + ((send_part % recv_part) > 0);
+    const int end_part = MPL_MIN(start_part + n_part, send_part);
+    MPIR_Assert(start_part >= 0);
+    MPIR_Assert(end_part >= -1);
+
+    // get the number of partitions that are still busy. if the sum is 0 then they have all finished
+    int is_busy = (too_early);
+    for (int ip = start_part; ip < end_part; ++ip) {
+        is_busy += MPIR_cc_get(cc_part[ip]);
+    }
```

Comment:
this is still an all or nothing check, right? does the main cc for the request not work in that case? obviously fine grained parrived support will need to check the `cc_part` array, but not seeing the use at this stage.

Reply:
no it's the fine-grained one. if the number of send and recv partitions are different you need to visit all the required partitions to answer the `Parrived`.
if I have `18` send partitions and `5` receive ones and I want to check the status of partition `i=2` for example, I need to loop over send partitions from `7` to `10` included. 
with this code I get `start_part=36/5=7` and `npart=3+1 = 4`

Reply:
Oh, I see it now. Disregard 😄.